lib/SIL/Utils/OwnershipUtils.cpp

```diff
@@ -2383,21 +2383,62 @@ bool swift::isNestedLexicalBeginBorrow(BeginBorrowInst *bbi) {
 }
 
 bool swift::isRedundantMoveValue(MoveValueInst *mvi) {
+  // Check whether the moved-from value's lifetime and the moved-to value's
+  // lifetime have the same (1) ownership, (2) lexicality, and (3) escaping.
+  //
+  // Along the way, also check for cases where they have different values for
+  // those characteristics but it doesn't matter because of how limited the uses
+  // of the original value are (for now, whether the move is the only use).
+
   auto original = mvi->getOperand();
 
-  // If the moved-from value has none ownership, hasPointerEscape can't handle
-  // it, so it can't be used to determine whether escaping matches.
+  // (1) Ownership matches?
+  // (The new value always has owned ownership.)
   if (original->getOwnershipKind() != OwnershipKind::Owned) {
     return false;
   }
 
-  // First, check whether lexicality matches, the cheaper check.
+  // (2) Lexicality matches?
   if (mvi->isLexical() != original->isLexical()) {
     return false;
   }
 
-  // Then, check whether escaping matches, the more expensive check.
-  if (hasPointerEscape(mvi) != hasPointerEscape(original)) {
+  // The move doesn't alter constraints: ownership and lexicality match.
+
+  auto *singleUser =
```

Comment:
The comments are great. But I still somehow missed that that singleUser check is actually part of condition #3, and only a way to short-circuit the complete, but slower, logic below.

Reply:
Follow-up at https://github.com/apple/swift/pull/64483 .